insta/src/snapshot.rs

```diff
@@ -511,42 +596,111 @@ impl Snapshot {
         self.save_with_metadata(&new_path, &self.metadata)?;
         Ok(new_path)
     }
+
+    pub fn cleanup_extra_files(snapshot: Option<&Self>, path: &Path) -> Result<(), Box<dyn Error>> {
```

Comment:
Why do we need this function? (I don't doubt we do, I just don't quite get grok it at first...)

Reply:
If the extension of a pending snapshot changes when re-running an unapproved test (or goes from a binary to a text snapshot) it would otherwise just write it with the new extension and overwrite the .snap. The alternative would be to try to read the pending file path of what is being asserted and then delete the file with the extension from in there. And then I also used it when approving or rejecting a pending snapshot, just because that's a bit simpler than using the extension to derive the path to delete (but in that case the metadata is already in memory).

This approach was more necessary with the other implementation where we created the file as a `._` first which also had to be cleaned up if a crash happened anywhere in between creating it at the beginning of `assert_snapshot` and where the `.snap` was actually being written.

But I still think this is the simpler approach compared to loading the previous pending snapshot and looking at the extension field there.

Reply:
Can we integrate with the existing cleanup functions — either `cleanup_passing` or removing unreferenced snapshots? 

Those file tree walkers that will find snapshots reliably too.

Reply:
If I understood this correctly, I think the unreferences snapshot removal happens fully in `cargo-insta`. I think this needs to happen when just running with `cargo test` as well, otherwise there can be a bunch of different files lying around.

To integrate it with `cleanup_passing` we'd have to change that into a general `cleanup` function that doesn't only run in the passing case. Also currently `cleanup_passing` only tries to delete one file: the one with the `.new` appended. And `cleanup_extra_files` currently relies on being run before the new files are stored, because otherwise it would delete the new pending snapshot.

I gave it a shot and it ended up looking like this:

```rust
pub fn cleanup(&self, passing: bool) -> Result<(), Box<dyn Error>> {
    if let Some(ref path) = self.snapshot_file {
        let binary_path =
            self.old_snapshot
                .as_ref()
                .and_then(|snapshot| match &snapshot.contents() {
                    SnapshotContents::Text(_) => None,
                    SnapshotContents::Binary(contents) => Some(contents.build_path(path)),
                });
        let binary_file_name = binary_path.as_ref().map(|path| path.file_name().unwrap());

        // The file name to compare against has to be valid utf-8 as it is generated by this
        // crate out of utf-8 strings.
        let file_name = path.file_name().unwrap().to_str().unwrap();

        let pending_path = path.with_extension("snap.new");
        let pending_file_name = pending_path.file_name().unwrap();

        let pending_binary_path =
            self.old_snapshot
                .as_ref()
                .and_then(|snapshot| match &snapshot.contents() {
                    SnapshotContents::Text(_) => None,
                    SnapshotContents::Binary(contents) => {
                        Some(contents.build_path(&pending_path))
                    }
                });
        let pending_binary_file_name = pending_binary_path
            .as_ref()
            .map(|path| path.file_name().unwrap());

        let read_dir = path.parent().unwrap().read_dir();

        match read_dir {
            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(()),
            _ => (),
        }

        for entry in read_dir? {
            let entry = entry?;
            let entry_file_name = entry.file_name();

            // We'll just skip over files with non-utf-8 names. The assumption being that those
            // would not have been generated by this crate.
            if entry_file_name
                .to_str()
                .map(|f| f.starts_with(file_name))
                .unwrap_or(false)
                && entry_file_name != file_name
                && binary_file_name.map_or(true, |x| x != entry_file_name)
                && (passing
                    || (entry_file_name != pending_file_name
                        && pending_binary_file_name.map_or(true, |x| x != entry_file_name)))
            {
                std::fs::remove_file(entry.path())?;
            }
        }
    }

    if passing {
        // and add a null pending snapshot to a pending snapshot file if needed
        if let Some(ref pending_snapshots) = self.pending_snapshots_path {
            if fs::metadata(pending_snapshots).is_ok() {
                PendingInlineSnapshot::new(None, None, self.assertion_line)
                    .save(pending_snapshots)?;
            }
        }
    }

    Ok(())
}
```

But then I realized that it's also used in `cargo-insta` in `Container::commit`. There we don't have an `AssertionContext` (and probably also shouldn't). So we'd need a different solution there.

Right now I can't really think of a way to integrate the two that doesn't increase the overall complexity.

Reply:
OK, I get some of this, though haven't quite groked why we need a separate path. Will have another look through later. 

If possible, I think we should try and integrate with the existing functions, so the semantics are the same and the internals remain simple. For example — while this isn't a huge deal — it looks like if we change the extension, we [remove unreferenced snapshots](https://github.com/mitsuhiko/insta/blob/654e1a6ee8038a616d254fd8376f4e199aab6780/cargo-insta/tests/main.rs#L1328) (nice tests!) — but we generally wouldn't do that (without a `--delete-unreferenced-snapshots` run)

For example, we already have the `try_removing_snapshot` on the pending path (https://github.com/mitsuhiko/insta/blob/8753f2cc59284bd8bc8c645594e524fd16495423/cargo-insta/src/container.rs#L231-L242) — could we instead extend that to also remove the a pending binary snapshot when it's accepted? (Maybe the answer to this is "not without making the code more complicated, because that just takes a path, and this would need to take an object which looked up the pending paths")

Reply:
Ok, I've changed some things around. This function was a little bit weird because it deleted all the files that start with the passed in path. While that was correct, it's a little bit confusing and I think the exact way I did it is related to there being an extra temporary file in the previous implementation.

I've now changed the function to only deal with any additional `.snap.new.*` files in with the same snapshot filename. This is the alternative to loading the previous pending snapshot and then deleting the binary snapshot from there. (We could also do that, but in that case I'd want to only load the metadata to prevent having an extra copy of the binary in-memory). In cargo-insta I've changed it to just delete the binaries there, because the snapshots are already loaded.

Reply:
Thank you!